### ISSUE TITLE
The whole application is shutdowned when using illegal parameter for query

### ISSUE THREAD

**Issue by johnzeng** [mysql-otp/mysql-otp#66]
```
(eLockAccount@johnzeng-virtual-machine)1> mysql:query(pool_mysql, <<"select * from account where id =?">>, [undefine]).

=ERROR REPORT==== 14-Apr-2017::11:05:01 ===
** Generic server <0.1261.0> terminating
** Last message in was {param_query,<<"select * from account where id =?">>,
                                    [undefine]}
** When Server state == {state,[5,5,54],
                               95,#Port<0.3137>,"localhost",3306,"root",
                               "Removed",true,infinity,infinity,60000,0,
                               2,0,0,0,undefined,
                               {dict,0,16,16,8,80,48,
                                     {[],[],[],[],[],[],[],[],[],[],[],[],[],
                                      [],[],[]},
                                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                                       [],[],[]}}},
                               empty,false}
** Reason for termination ==
** {function_clause,[{mysql_protocol,encode_param,
                                     [undefine],
                                     [{file,"src/mysql_protocol.erl"},
                                      {line,745}]},
                     {lists,map,2,[{file,"lists.erl"},{line,1238}]},
                     {mysql_protocol,execute,5,
                                     [{file,"src/mysql_protocol.erl"},
                                      {line,161}]},
                     {mysql,execute_stmt,4,
                            [{file,"src/mysql.erl"},{line,804}]},
                     {gen_server,try_handle_call,4,
                                 [{file,"gen_server.erl"},{line,629}]},
                     {gen_server,handle_msg,5,
                                 [{file,"gen_server.erl"},{line,661}]},
                     {proc_lib,init_p_do_apply,3,
                               [{file,"proc_lib.erl"},{line,240}]}]}
** exception exit: {{function_clause,[{mysql_protocol,encode_param,
                                                      [undefine],
                                                      [{file,"src/mysql_protocol.erl"},{line,745}]},
                                      {lists,map,2,[{file,"lists.erl"},{line,1238}]},
                                      {mysql_protocol,execute,5,
                                                      [{file,"src/mysql_protocol.erl"},{line,161}]},
                                      {mysql,execute_stmt,4,[{file,"src/mysql.erl"},{line,804}]},
                                      {gen_server,try_handle_call,4,
                                                  [{file,"gen_server.erl"},{line,629}]},
                                      {gen_server,handle_msg,5,
                                                  [{file,"gen_server.erl"},{line,661}]},
                                      {proc_lib,init_p_do_apply,3,
                                                [{file,"proc_lib.erl"},{line,240}]}]},
                    {gen_server,call,
                                [pool_mysql,
                                 {param_query,<<"select * from account where id =?">>,
                                              [undefine]},
                                 infinity]}}
     in function  gen_server:call/3 (gen_server.erl, line 212)
     in call from mysql:query_call/2 (src/mysql.erl, line 792)
(eLockAccount@johnzeng-virtual-machine)2>
=INFO REPORT==== 14-Apr-2017::11:05:01 ===
    application: eLockAccount
    exited: shutdown
    type: permanent
{"Kernel pid terminated",application_controller,"{application_terminated,eLockAccount,shutdown}"}

Crash dump is being written to: erl_crash.dump...done
Kernel pid terminated (application_controller) ({application_terminated,eLockAccount,shutdown})
make: *** [console] Error 1
```

Is it possilbe to add param check to avoid it?? This it unacceptable for an erlang application.

I did some search on it and looks like the crash comes from the `mysql:terminate/2` callback.

I modified the callback to do some logging:

```
terminate(Reason, #state{socket = Socket})
  when Reason == normal; Reason == shutdown ->
      io:format("now shutdown for reason:~p~n", [Reason]),
      %% Send the goodbye message for politeness.
      inet:setopts(Socket, [{active, false}]),
      io:format("setopts is done~n"),
      R = mysql_protocol:quit(gen_tcp, Socket),
      io:format("quit gen_tcp is done~n"),
      inet:setopts(Socket, [{active, once}]),
      io:format("set opts done again~n"),
      R;
```

and the output is like this:

```
now shutdown for reason:shutdown
(eLockAccount@johnzeng-virtual-machine)2>
=INFO REPORT==== 14-Apr-2017::14:56:17 ===
    application: eLockAccount
    exited: shutdown
    type: permanent
{"Kernel pid terminated",application_controller,"{application_terminated,eLockAccount,shutdown}"}
```

So the exception happens when calling 
```
inet:setopts(Socket, [{active, false}]),
```

Please let me know what's wrong here. I don't wanna dig into it anymore.

**Comment by zuiderkwast** [comment]
In the query `mysql:query(pool_mysql, <<"select * from account where id =?">>, [undefine]).`, is `pool_mysql` a connection or a pool of some sort? If it is a pool, it will definitely not work.

**Comment by zuiderkwast** [comment]
If the connection process crashes, `mysql:query(...)` will raise an exception. If you don't catch that exception, it will crash the caller process as well.

The error in the terminate function doesn't make a difference. (The socket is probably already closed when we try to set `{active, false}`.) Terminate runs in the gen_server process which crashes anyway. (We could perhaps ignore the errors in terminate by catching them, but I don't think it matters.)

**Comment by johnzeng** [comment]
pool_mysql is not pool name , just connection name.

**Comment by zuiderkwast** [comment]
Ok, the problem is the parameter. It is possible to add parameter check and raise an error {badarg, undefine}, just like many erlang functuons do when you pass invalid arguments. Then you can catch the error without crashing the connection.

You say "This it unacceptable for an erlang application" but actually the erlang philosophy is "let it crash". In many other languages I would agree this is unacceptable. Maybe in erlang too this would make sense, but it would make the implementation more bloated. But maybe you are right. I have not decided what i think is best :)

**Comment by zuiderkwast** [comment]
Any updates on this? I would accept a PR where connection process catches the error and returns something to the the query function, which raises a badarg error in the caller process. Or if we can validate the parameters before passing them to the connection process, that would work too.

**Comment by juhlig** [comment]
> You say "This it unacceptable for an erlang application" but actually the erlang philosophy is "let it crash".

I at once agree and disagree with both of you. I think it should crash the _caller_ (admittedly, in a sense it does ;)), ie the one responsible for giving invalid input, but not the "innocent" recipient of invalid input. That should (arguably) crash if something goes unexpectedly wrong when processing otherwise valid input.

> Any updates on this? I would accept a PR where connection process catches the error and returns something to the the query function, which raises a badarg error in the caller process. Or if we can validate the parameters before passing them to the connection process, that would work too.

I'll see to it. I think I'll go for validating before passing to the connection, it's simpler and seems more clear to me.

**Comment by zuiderkwast** [comment]
Solved by #113. Thanks @juhlig!
